File: Medium/Buy and Sell a Share at most twice/buy-and-sell-a-share-at-most-twice.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;


class IntArray
{
    public static int[] input(BufferedReader br, int n) throws IOException
    {
        String[] s = br.readLine().trim().split(" ");
        int[] a = new int[n];
        for(int i = 0; i < n; i++)
            a[i] = Integer.parseInt(s[i]);
        
        return a;
    }
    
    public static void print(int[] a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }
    
    public static void print(ArrayList<Integer> a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine().trim());
        while(t-- > 0){
            
            int n; 
            n = Integer.parseInt(br.readLine().trim());
            
            
            int[] price = IntArray.input(br, n);
            
            Solution obj = new Solution();
            int res = obj.maxProfit(n, price);
            
            System.out.println(res);
            
        }
    }
}

// } Driver Code Ends


class Solution {
    public static int maxProfit(int n, int[] price) {
        // code here
        int[] dp1 = new int[n];
        int[] dp2 = new int[n];
        int least = price[0];
        for(int i = 1; i<n; i++) {
            least = Math.min(least, price[i]);
            int val = price[i] - least;
            dp1[i] = Math.max(dp1[i-1], val);
        }
        
        int highest = price[n-1];
        for(int i = n-2; i>=0; i--) {
            highest = Math.max(highest, price[i]);
            int val = highest - price[i];
            dp2[i] = Math.max(dp2[i+1], val);
        }
        
        int ans = 0;
        for(int i = 0; i<n; i++) {
            ans = Math.max(ans, dp1[i] + dp2[i]);
        }
        
        return ans;
    }
}
        

File: Medium/Buy and Sell a Share at most twice/buy-and-sell-a-share-at-most-twice.py
from typing import List


class Solution:
    def maxProfit(self, n : int, price : List[int]) -> int:
        # code here
        dp1, dp2 = [0] * n, [0] * n
        least, highest = price[0], price[-1]
        
        for i in range(1, n):
            least = min(price[i], least)
            val = price[i] - least
            dp1[i] = max(dp1[i-1], val);
            
        for i in range(n-2, -1, -1):
            highest = max(highest, price[i])
            val = highest - price[i]
            dp2[i] = max(dp2[i+1], val)
            
        ans = 0
        for i in range(n):
            ans = max(ans, dp1[i] + dp2[i])
        return ans
            
         



#{ 
 # Driver Code Starts
class IntArray:
    def __init__(self) -> None:
        pass
    def Input(self,n):
        arr=[int(i) for i in input().strip().split()]#array input
        return arr
    def Print(self,arr):
        for i in arr:
            print(i,end=" ")
        print()


if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        
        n = int(input())
        
        
        price=IntArray().Input(n)
        
        obj = Solution()
        res = obj.maxProfit(n, price)
        
        print(res)
        

# } Driver Code Ends